src/app/ReadClient.cpp

```diff
@@ -0,0 +1,283 @@
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines read client for a CHIP Interaction Data model
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/ReadClient.h>
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR ReadClient::Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    // Error if already initialized.
+    VerifyOrExit(mpExchangeMgr == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(mpExchangeCtx == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    mpExchangeMgr = apExchangeMgr;
+    mpExchangeCtx = nullptr;
+    mpDelegate    = apDelegate;
+    mState        = ClientState::Initialized;
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadClient::Shutdown()
+{
+    ClearExistingExchangeContext();
+    mpExchangeMgr = nullptr;
+    mpDelegate    = nullptr;
+    MoveToState(ClientState::Uninitialized);
+}
+
+void ReadClient::Reset()
+{
+    ClearExistingExchangeContext();
+    MoveToState(ClientState::Initialized);
+}
+
+const char * ReadClient::GetStateStr() const
+{
+#if CHIP_DETAIL_LOGGING
+    switch (mState)
+    {
+    case ClientState::Uninitialized:
+        return "UNINIT";
+    case ClientState::Initialized:
+        return "INIT";
+    case ClientState::ReadRequestSending:
+        return "READREQUESTSENDING";
+    }
+#endif // CHIP_DETAIL_LOGGING
+    return "N/A";
+}
+
+void ReadClient::MoveToState(const ClientState aTargetState)
+{
+    mState = aTargetState;
+    ChipLogDetail(DataManagement, "Client[%u] moving to [%5.5s]",
+                  InteractionModelEngine::GetInstance()->GetReadClientArrayIndex(this), GetStateStr());
+}
+
+void ReadClient::ClearState(void)
+{
+    MoveToState(ClientState::Uninitialized);
+}
+
+CHIP_ERROR ReadClient::SendReadRequest(NodeId aNodeId, Transport::AdminId aAdminId)
+{
+    CHIP_ERROR err                    = CHIP_NO_ERROR;
+    System::PacketBufferHandle msgBuf = nullptr;
+    chip::app::InteractionModelDelegate::InParam inParam;
+    chip::app::InteractionModelDelegate::OutParam outParam;
+
+    ChipLogDetail(DataManagement, "Client[%u] [%5.5s]", InteractionModelEngine::GetInstance()->GetReadClientArrayIndex(this),
+                  GetStateStr());
+    VerifyOrExit(ClientState::Initialized == mState, err = CHIP_ERROR_INCORRECT_STATE);
+
+    if (nullptr != mpDelegate)
+    {
+        outParam.mReadRequestPrepareNeeded.eventPathParamsList = nullptr;
+        mpDelegate->HandleIMCallBack(chip::app::InteractionModelDelegate::CallbackId::kReadRequestPrepareNeeded, inParam, outParam);
+    }
+
+    {
+        chip::System::PacketBufferTLVWriter writer;
+        ReadRequest::Builder request;
+
+        msgBuf = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLength);
+        VerifyOrExit(!msgBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+        writer.Init(std::move(msgBuf));
```

Comment:
you move msgBuf here then continue to use it below. This does not look safe. Why is it ok?

Reply:
I suggest to stop using the `goto exit` pattern and use return code instead. It is hard to follow object lifecycles when we are forced to declare all objects at the top of the scope because of initialization order.